# **`Dir e Help no Python`**

##### **dir** - Apresenta todos os atributos/Propriedades e funções/metodos disponiveis para determinado tipo de dado ou variavel.

##### **help** - Apresenta a documentação/como utilizar os atributos/propriedades e funções/metodos disponiveis para determinado tipo de dado ou variavel.

Exatamente! No exemplo dado:

```python
nome = 'Francisco Douglas'
help(nome) # Não mostrará nada nesse caso, pois 'nome' é uma string e não um objeto que possui documentação
dir(nome) # Irá mostrar todos os métodos e atributos disponíveis para strings
```

O comando `help(nome)` não mostrará nada porque `nome` é uma string, e strings em Python são objetos simples sem documentação disponível diretamente através do comando `help()`. No entanto, você pode usar o comando `dir()` para listar todos os métodos e atributos disponíveis para a string `nome`. Isso pode ser útil para explorar as funcionalidades disponíveis para manipulação de strings em Python.

In [2]:
#Exemplo
nome = 'Francisco Douglas'
help(nome) #Não Mostrara nada nesse caso devo utiliza o dir
dir(nome) # Mim mostrara todos os metodos e funções

No Python documentation found for 'Francisco Douglas'.
Use help() to get the interactive help utility.
Use help(str) for help on the str class.



['__add__',
 '__class__',
 '__contains__',
 '__delattr__',
 '__dir__',
 '__doc__',
 '__eq__',
 '__format__',
 '__ge__',
 '__getattribute__',
 '__getitem__',
 '__getnewargs__',
 '__getstate__',
 '__gt__',
 '__hash__',
 '__init__',
 '__init_subclass__',
 '__iter__',
 '__le__',
 '__len__',
 '__lt__',
 '__mod__',
 '__mul__',
 '__ne__',
 '__new__',
 '__reduce__',
 '__reduce_ex__',
 '__repr__',
 '__rmod__',
 '__rmul__',
 '__setattr__',
 '__sizeof__',
 '__str__',
 '__subclasshook__',
 'capitalize',
 'casefold',
 'center',
 'count',
 'encode',
 'endswith',
 'expandtabs',
 'find',
 'format',
 'format_map',
 'index',
 'isalnum',
 'isalpha',
 'isascii',
 'isdecimal',
 'isdigit',
 'isidentifier',
 'islower',
 'isnumeric',
 'isprintable',
 'isspace',
 'istitle',
 'isupper',
 'join',
 'ljust',
 'lower',
 'lstrip',
 'maketrans',
 'partition',
 'removeprefix',
 'removesuffix',
 'replace',
 'rfind',
 'rindex',
 'rjust',
 'rpartition',
 'rsplit',
 'rstrip',
 'split',
 'splitlines',
 'startswith',
 'stri

Ao utilizar o comando `help()` com métodos de string específicos, como `nome.title`, `nome.lower` e `nome.upper`, você pode acessar a documentação embutida do Python para esses métodos. Aqui está um exemplo de como isso pode ser feito:

```python
# Documentação para o método title()
help(nome.title)
print('-'*80)

# Documentação para o método lower()
help(nome.lower)
print('*'*80)

# Documentação para o método upper()
help(nome.upper)
```

Isso mostrará a documentação embutida do Python para os métodos `title()`, `lower()` e `upper()` respectivamente. Esta documentação fornecerá uma descrição do que cada método faz e como usá-lo, juntamente com informações adicionais, como parâmetros e valores de retorno.

In [4]:
#Exemplo documentacao title
nome = 'francisco'
help(nome.title)
print('-'*80)

#documentacao lower
help(nome.lower)
print('*'*80)

#documentacao upper
help(nome.upper)

Help on built-in function title:

title() method of builtins.str instance
    Return a version of the string where each word is titlecased.
    
    More specifically, words start with uppercased characters and all remaining
    cased characters have lower case.

--------------------------------------------------------------------------------
Help on built-in function lower:

lower() method of builtins.str instance
    Return a copy of the string converted to lowercase.

********************************************************************************
Help on built-in function upper:

upper() method of builtins.str instance
    Return a copy of the string converted to uppercase.



Excelente! Aqui está o código usando `dir()` para listar os métodos e atributos disponíveis para diferentes tipos de dados:

```python
inteiro = 23
fracionario = 45.67
strings = 'Programador Python'
boleano = True

# Métodos e atributos para números inteiros
print(f'Métodos do Inteiro: {dir(inteiro)}\n')

# Métodos e atributos para números de ponto flutuante
print(f'Métodos do Float: {dir(fracionario)}\n')

# Métodos e atributos para strings
print(f'Métodos da String: {dir(strings)}\n')

# Métodos e atributos para valores booleanos
print(f'Métodos do Booleano: {dir(boleano)}')
```

Isso mostrará todos os métodos e atributos disponíveis para cada tipo de dado. Você verá uma lista de métodos e atributos específicos para cada tipo de dado, como inteiros, floats, strings e booleanos. Essa é uma maneira útil de explorar as funcionalidades disponíveis para diferentes tipos de dados em Python.

In [5]:
#Utilizando o Dir
inteiro     = 23
fracionario = 45.67
strings     = 'Programador Python'
doubles     = True

#Metodos e Funções
print(f'Metodos do Int: {dir(inteiro)}\n')
print(f'Metodos do float: {dir(fracionario)}\n')
print(f'Metodos String: {dir(strings)}\n')
print(f'Metodos Booleano: {dir(doubles)}')

Metodos do Int: ['__abs__', '__add__', '__and__', '__bool__', '__ceil__', '__class__', '__delattr__', '__dir__', '__divmod__', '__doc__', '__eq__', '__float__', '__floor__', '__floordiv__', '__format__', '__ge__', '__getattribute__', '__getnewargs__', '__getstate__', '__gt__', '__hash__', '__index__', '__init__', '__init_subclass__', '__int__', '__invert__', '__le__', '__lshift__', '__lt__', '__mod__', '__mul__', '__ne__', '__neg__', '__new__', '__or__', '__pos__', '__pow__', '__radd__', '__rand__', '__rdivmod__', '__reduce__', '__reduce_ex__', '__repr__', '__rfloordiv__', '__rlshift__', '__rmod__', '__rmul__', '__ror__', '__round__', '__rpow__', '__rrshift__', '__rshift__', '__rsub__', '__rtruediv__', '__rxor__', '__setattr__', '__sizeof__', '__str__', '__sub__', '__subclasshook__', '__truediv__', '__trunc__', '__xor__', 'as_integer_ratio', 'bit_count', 'bit_length', 'conjugate', 'denominator', 'from_bytes', 'imag', 'numerator', 'real', 'to_bytes']

Metodos do float: ['__abs__', '__ad

## **Recebendo dados do usuario**

```python
# Imprime uma mensagem na tela
print('O print imprime na tela')

# Declarando uma variável e imprimindo seu valor
dados = 'Acabei de declarar uma variável'
print(dados)

# Solicita ao usuário que insira dados e imprime o valor inserido
dados2 = input('O input recebe os dados a serem digitados: ')
print(dados2)

# Formas de imprimir na tela usando f-string e format
print('\n')
print(f'Utilizando f-string: {dados2}')
print('Utilizando format: {}'.format(dados2))
```

**Explicação:**

1. **Imprimindo na tela:** O comando `print()` é usado para exibir mensagens na tela.

2. **Declarando variáveis:** O código declara uma variável chamada `dados` e atribui a ela a string `'Acabei de declarar uma variável'`. Em seguida, imprime o valor da variável.

3. **Recebendo dados do usuário:** O código usa a função `input()` para solicitar que o usuário insira dados. Os dados inseridos são armazenados na variável `dados2`, que é então impressa na tela.

4. **Formas de impressão na tela:** O código demonstra duas maneiras de imprimir na tela usando formatação de strings. A primeira usa f-strings, que são strings prefixadas com `f` e permitem inserir valores de variáveis diretamente na string. A segunda usa o método `.format()`, que substitui os marcadores de posição na string pelos valores das variáveis especificadas.

In [6]:
print('O print imprime na tela')

#Declarando variaveis
dados = 'Acabei de Declara uma Variavel'
print(dados)

#Recebendo dados do usuario
dados2 = input('O Input recebe os dados  serem digitados: ')
print(dados2)

#Formas de imprimir na tela o melhores que acho
print('\n')
print(f'Utilizando f,string: {dados2}')
print('Utilizando f, format {}'.format(dados2))

O print imprime na tela
Acabei de Declara uma Variavel
der


Utilizando f,string: der
Utilizando f, format der


# **`TIPOS DE DADOS EM PYTHON`**


```python
# Tipo de dado: int (inteiro)
# Descrição: Representa números inteiros, como 2, 3, 4, etc.
# Exemplo:
inteiro = 10

# Tipo de dado: float (ponto flutuante)
# Descrição: Representa números de ponto flutuante, como 4.3, 5.667, 2.23, etc.
# Exemplo:
flutuante = 3.14

# Tipo de dado: bool (booleano)
# Descrição: Representa valores lógicos Verdadeiro (True) e Falso (False).
# Exemplo:
booleano = True

# Tipo de dado: str (string)
# Descrição: Representa texto, palavras e tudo o que estiver entre aspas. 
# Exemplo:
texto = 'Olá, mundo!'
```

**Explicação:**

1. **int (inteiro):** Este tipo de dado representa números inteiros, como 2, 3, 4, etc.

2. **float (ponto flutuante):** Este tipo de dado representa números de ponto flutuante, como 4.3, 5.667, 2.23, etc.

3. **bool (booleano):** Este tipo de dado representa valores lógicos Verdadeiro (True) e Falso (False).

4. **str (string):** Este tipo de dado representa texto, palavras e tudo o que estiver entre aspas (simples ou duplas). Por exemplo, `'Olá, mundo!'` é uma string.

### Tipo int

In [7]:
#Tipo int
numero_int = 2
print(f'Numero inteiro {numero_int} {type(numero_int)}')

Numero inteiro 2 <class 'int'>


### Tipo float

In [8]:
#Tipo float
num_float = 45.6
print(f'Numero Float {num_float} {type(num_float)}')

Numero Float 45.6 <class 'float'>


 `int` e `float`:

```python
# Contagem de métodos e funções para o tipo de dado int
print(f'Ao todo temos: {len(dir(int))} métodos/funções do tipo int')

# Contagem de métodos e funções para o tipo de dado float
print(f'Ao todo temos: {len(dir(float))} métodos/funções do tipo float')
```

Este código mostrará quantos métodos e funções estão disponíveis para cada tipo de dado. Esses métodos e funções são as operações que podem ser realizadas com esses tipos de dados em Python.

In [9]:
#Metodos e funcoes do float e int
print(f'Ao todo temos: {len(dir(int))} metodos/funcoes do tipo int')
print(f'Ao todo temos: {len(dir(float))} metodos/funcoes do tipo float')

Ao todo temos: 73 metodos/funcoes do tipo int
Ao todo temos: 59 metodos/funcoes do tipo float


In [10]:
#Print do int
dir(int)
help(int)

Help on class int in module builtins:

class int(object)
 |  int([x]) -> integer
 |  int(x, base=10) -> integer
 |  
 |  Convert a number or string to an integer, or return 0 if no arguments
 |  are given.  If x is a number, return x.__int__().  For floating point
 |  numbers, this truncates towards zero.
 |  
 |  If x is not a number or if base is given, then x must be a string,
 |  bytes, or bytearray instance representing an integer literal in the
 |  given base.  The literal can be preceded by '+' or '-' and be surrounded
 |  by whitespace.  The base defaults to 10.  Valid bases are 0 and 2-36.
 |  Base 0 means to interpret the base from the string as an integer literal.
 |  >>> int('0b100', base=0)
 |  4
 |  
 |  Built-in subclasses:
 |      bool
 |  
 |  Methods defined here:
 |  
 |  __abs__(self, /)
 |      abs(self)
 |  
 |  __add__(self, value, /)
 |      Return self+value.
 |  
 |  __and__(self, value, /)
 |      Return self&value.
 |  
 |  __bool__(self, /)
 |      True if 

In [9]:
#float
dir(float)
help(float)

Help on class float in module builtins:

class float(object)
 |  float(x=0, /)
 |  
 |  Convert a string or number to a floating point number, if possible.
 |  
 |  Methods defined here:
 |  
 |  __abs__(self, /)
 |      abs(self)
 |  
 |  __add__(self, value, /)
 |      Return self+value.
 |  
 |  __bool__(self, /)
 |      True if self else False
 |  
 |  __ceil__(self, /)
 |      Return the ceiling as an Integral.
 |  
 |  __divmod__(self, value, /)
 |      Return divmod(self, value).
 |  
 |  __eq__(self, value, /)
 |      Return self==value.
 |  
 |  __float__(self, /)
 |      float(self)
 |  
 |  __floor__(self, /)
 |      Return the floor as an Integral.
 |  
 |  __floordiv__(self, value, /)
 |      Return self//value.
 |  
 |  __format__(self, format_spec, /)
 |      Formats the float according to format_spec.
 |  
 |  __ge__(self, value, /)
 |      Return self>=value.
 |  
 |  __getattribute__(self, name, /)
 |      Return getattr(self, name).
 |  
 |  __getnewargs__(self, /)
 

### **Tipo Booleano bool so aceita True ou False**

In [11]:
#Tipo bool
verdade = True
falso   = False

if verdade == falso:
    print(verdade)
else:
    print(falso)

#Sao condicoes logicas

False


In [12]:
#Metodos e documentacao
dir(bool)
help(bool)

Help on class bool in module builtins:

class bool(int)
 |  bool(x) -> bool
 |  
 |  Returns True when the argument x is true, False otherwise.
 |  The builtins True and False are the only two instances of the class bool.
 |  The class bool is a subclass of the class int, and cannot be subclassed.
 |  
 |  Method resolution order:
 |      bool
 |      int
 |      object
 |  
 |  Methods defined here:
 |  
 |  __and__(self, value, /)
 |      Return self&value.
 |  
 |  __or__(self, value, /)
 |      Return self|value.
 |  
 |  __rand__(self, value, /)
 |      Return value&self.
 |  
 |  __repr__(self, /)
 |      Return repr(self).
 |  
 |  __ror__(self, value, /)
 |      Return value|self.
 |  
 |  __rxor__(self, value, /)
 |      Return value^self.
 |  
 |  __xor__(self, value, /)
 |      Return self^value.
 |  
 |  ----------------------------------------------------------------------
 |  Static methods defined here:
 |  
 |  __new__(*args, **kwargs) from builtins.type
 |      Create 

### Tipo string str

In [13]:
#str
nome = 'Palavras, textos, "45.56", Numeros entre aspas'
print(f'String {nome} = {type(nome)}')

String Palavras, textos, "45.56", Numeros entre aspas = <class 'str'>


In [14]:
# String , tudo que estive emtre aspas simples '', aspas duplas "" e aspas triplas """"""
# Ou seja tudo que estive entre aspas e string

nome = 'Francisco Douglas'
print(f"Seu nome e {nome} 'Voce e um programador'")

print(f"""
      Utilizando aspas Triplas
      Tambem utilizando aspas duplas dentro de simples "OK Python",
      Agora utilizando aspas simples 'Aprendendo Python'
      {nome}
      """)

Seu nome e Francisco Douglas 'Voce e um programador'

      Utilizando aspas Triplas
      Tambem utilizando aspas duplas dentro de simples "OK Python",
      Agora utilizando aspas simples 'Aprendendo Python'
      Francisco Douglas
      


In [15]:
#Metodos documentacao
dir(str)
help(str)

Help on class str in module builtins:

class str(object)
 |  str(object='') -> str
 |  str(bytes_or_buffer[, encoding[, errors]]) -> str
 |  
 |  Create a new string object from the given object. If encoding or
 |  errors is specified, then the object must expose a data buffer
 |  that will be decoded using the given encoding and error handler.
 |  Otherwise, returns the result of object.__str__() (if defined)
 |  or repr(object).
 |  encoding defaults to sys.getdefaultencoding().
 |  errors defaults to 'strict'.
 |  
 |  Methods defined here:
 |  
 |  __add__(self, value, /)
 |      Return self+value.
 |  
 |  __contains__(self, key, /)
 |      Return key in self.
 |  
 |  __eq__(self, value, /)
 |      Return self==value.
 |  
 |  __format__(self, format_spec, /)
 |      Return a formatted version of the string as described by format_spec.
 |  
 |  __ge__(self, value, /)
 |      Return self>=value.
 |  
 |  __getattribute__(self, name, /)
 |      Return getattr(self, name).
 |  
 |  

## Conversão de dados

```python
# Definição de variáveis
n_int = 45          # Número inteiro
n_float = 34.45     # Número float
n_string = 'Uma String' # String
n_bool = True       # Booleano

# Convertendo de int para float
de_int_float = float(n_int)
print(f'{de_int_float} {type(de_int_float)}')

# Convertendo de int para string
de_int_str = str(n_int)
print(f'{de_int_str} {type(de_int_str)}')

# Convertendo de int para booleano
de_int_bool = bool(n_int)
print(de_int_bool)
```

**Explicação:**

1. **Convertendo de int para float:** Usando a função `float()`, podemos converter um número inteiro em um número de ponto flutuante.

2. **Convertendo de int para string:** Usando a função `str()`, podemos converter um número inteiro em uma string.

3. **Convertendo de int para booleano:** Em Python, qualquer número diferente de zero é considerado verdadeiro (True) quando convertido para um booleano. Portanto, ao converter um número inteiro para booleano, se o número for diferente de zero, o valor será True; caso contrário, será False. No caso do número inteiro 45, que é diferente de zero, a saída será True.

In [17]:
#Podemos converter os dados e virse versa
n_int = 45 #Numero inteiro
n_float = 34.45 #Numero float
n_string = 'Uma String'
n_bool = True #Booleano

#Convertendo
de_int_float = float(n_int) #de inteiro para float
print(f'{de_int_float} {type(de_int_float)}')

de_int_str = str(n_int) #de inteiro a string
print(f'{de_int_str} {type(de_int_str)}')

#Posso converter qualquer tipo de dado e virse versa

de_int_bool = bool(n_int)
print(de_int_bool)

45.0 <class 'float'>
45 <class 'str'>
True




```python
# Definição de uma variável global chamada 'variavel_global'
variavel_global = 'Isso é considerado uma variável global'

# Loop for para demonstrar uma variável local
for n in range(1, 11):
    # Dentro do loop, definimos uma variável local chamada 'abc'
    abc = 'ABC será uma variável local'

# Tentativa de imprimir a variável local 'abc' fora do loop
print(abc)
```

**Explicação:**

1. **Variável Global:** Nós declaramos uma variável chamada `variavel_global` fora de qualquer função ou loop, o que a torna acessível em todo o programa.

2. **Loop For:** Usamos um loop `for` para iterar de 1 a 10. Dentro deste loop, definimos uma variável chamada `abc`. Esta variável é uma variável local, o que significa que só pode ser acessada dentro do loop.

3. **Tentativa de Impressão:** Tentamos imprimir a variável `abc` fora do loop. No entanto, isso resultará em um erro porque `abc` é uma variável local e não está mais disponível fora do escopo do loop.

In [19]:
# Variaveis Globais
variavel_global = 'Isso e considerado uma variavel global'

#Variavel local
for n in range(1, 11):
    abc = 'ABC sera uma variavel local'

print(abc)

ABC sera uma variavel local


# **Essenciais**

In [20]:
#Separador sep
#separa de acordo que o que eu escolher
print(12, 14, "Meu nome", True, sep=" %% ")


12 %% 14 %% Meu nome %% True


O parâmetro `sep` é usado para definir o separador entre os argumentos passados para a função `print()`. Aqui está uma explicação do seu código:

```python
# Usando o parâmetro sep para definir um separador personalizado
print(12, 14, "Meu nome", True, sep=" %% ")
```

**Explicação:**

- `12`, `14`, `"Meu nome"`, e `True` são os valores passados como argumentos para a função `print()`.
- O parâmetro `sep=" %% "` define o separador entre esses valores como `" %% "`. Isso significa que entre cada valor impresso, haverá dois sinais de porcentagem com espaços em ambos os lados, resultando em uma saída como `12 %% 14 %% Meu nome %% True`.

In [11]:
#Quebra de linha
# \r e \n
# \r - retorno
# \n - quebra linha

print("Novo mundo esta aqui \nprint\n")
print("Novo Mundo \rPPoo")
print("Aqui embaixo")

Novo mundo esta aqui 
print

PPoo Mundo 
Aqui embaixo


Certo! Aqui está a explicação do seu código:

```python
# Usando \n para quebrar linha
print("Novo mundo esta aqui \nprint\n")

# Usando \r para retornar ao início da linha
print("Novo Mundo \rPPoo")

# Texto impresso após o uso de \r
print("Aqui embaixo")
```

**Explicação:**

- `\n`: É um caractere especial que representa uma quebra de linha. Quando utilizado em uma string dentro da função `print()`, ele faz com que o texto seguinte seja impresso na próxima linha.
- `\r`: É um caractere especial que representa um retorno ao início da linha. Quando utilizado em uma string dentro da função `print()`, ele faz com que o texto impresso a partir desse ponto volte para o início da linha atual, sobrescrevendo os caracteres anteriores.
- Quando usamos `\n` na primeira impressão, ele cria uma quebra de linha após o texto `"Novo mundo esta aqui "`, resultando em `"print"` sendo impresso em uma nova linha.
- Na segunda impressão, usamos `\r`, que faz com que `"PPoo"` seja impresso no lugar de `"Novo Mundo "`, resultando na saída `"Novo Mundo PPoo"`. A parte restante da string `"Aqui embaixo"` é impressa após essa linha.


In [22]:
#Utilizando o end
print('Meu nome e Francisco: ',end=' ')
print("Programador python")

Meu nome e Francisco:  Programador python


O parâmetro `end` é usado para especificar o que deve ser impresso no final da função `print()`.

```python
# Utilizando o parâmetro end para especificar o que deve ser impresso no final da função print()
print('Meu nome é Francisco: ', end=' ')

# Continua na mesma linha após o nome
print("Programador Python")
```

**Explicação:**

- `'Meu nome é Francisco: '` é a primeira string impressa pela função `print()`.
- `end=' '` especifica que após a primeira string, nada deve ser impresso, ou seja, a próxima impressão continuará na mesma linha.
- `"Programador Python"` é a segunda string impressa. Como o parâmetro `end` foi definido como um espaço em branco, esta string é impressa na mesma linha que a primeira.

## **Formatação de Strings**

In [1]:
"""
Formatacao basica de strings
s - string
d - int
f - float

.<numero de digitos>f
x ou X - Hexadecimal

(Caractere) (><^)(quantidade)
> - Esquerda
< - Direita
^ - Centro

Sinal - + ou -
Ex: 0>-100,.1f
Conversion flags - !r !s !a
"""

variavel = "ABC"
print(f'{variavel}')
print(f'{variavel:>10}')
print(f'{variavel:<10}')
print(f'{variavel:^10}')

#Numero
print(f'{192.38833:.1f}')


ABC
       ABC
ABC       
   ABC    
192.4




```python
# Definição de uma string
variavel = "ABC"

# Imprime a string sem formatação
print(f'{variavel}')

# Imprime a string alinhada à direita em um campo de 10 caracteres
print(f'{variavel:>10}')

# Imprime a string alinhada à esquerda em um campo de 10 caracteres
print(f'{variavel:<10}')

# Imprime a string centralizada em um campo de 10 caracteres
print(f'{variavel:^10}')

# Formata um número de ponto flutuante com uma casa decimal
print(f'{192.38833:.1f}')
```

**Explicação:**

- As f-strings (strings formatadas) permitem inserir expressões Python dentro de strings, facilitando a formatação.
- O operador `>` alinha o texto à direita, `<` alinha à esquerda e `^` centraliza.
- O número após o `:` em `{192.38833:.1f}` especifica que o número de ponto flutuante terá uma casa decimal.
- Isso resulta na impressão de `192.4`, arredondando o número para uma casa decimal.
- A formatação de strings é uma técnica poderosa em Python para exibir dados de forma legível e organizada.

In [2]:
# Fatiamento de String
variavel = "Estou programando em Python"
print(f'Tamanho da String: {len(variavel)}')
print(variavel[0: len(variavel) :4]) #Vai do 0 ate o tamanho da variavel pulando de 4 em 4
print(variavel[:12])    #Vai do 0 ate o 12
print(variavel[0:-1:3]) #Vai do 0 ate a ultima letra pulando de 3 em 3
print(variavel[::-1])   #Inverte toda a string
print(variavel.upper()) #Deixa tudo em Maiusculo
print(variavel.title()) #Deixa apenas a primeira letra de cada palavra em Maiuscula
print(variavel.lower()) #Deixa tudo em Minusculas

Tamanho da String: 27
Euomo h
Estou progra
EopgmdePh
nohtyP me odnamargorp uotsE
ESTOU PROGRAMANDO EM PYTHON
Estou Programando Em Python
estou programando em python




```python
# Definição da string
variavel = "Estou programando em Python"

# Imprime o tamanho da string
print(f'Tamanho da String: {len(variavel)}')

# Fatiamento com intervalo de 4 em 4, começando do primeiro caractere (índice 0)
print(variavel[0:len(variavel):4])

# Fatiamento do início até o índice 12 (exclusivo)
print(variavel[:12])

# Fatiamento do início até o penúltimo caractere, com intervalo de 3 em 3
print(variavel[0:-1:3])

# Inverte toda a string
print(variavel[::-1])

# Converte toda a string para maiúsculas
print(variavel.upper())

# Converte toda a string para o estilo de título (a primeira letra de cada palavra em maiúscula)
print(variavel.title())

# Converte toda a string para minúsculas
print(variavel.lower())
```

**Explicação:**

- **Fatiamento de strings:** O fatiamento de strings é realizado usando a sintaxe `[start:end:step]`, onde `start` é o índice inicial, `end` é o índice final (não incluso) e `step` é o tamanho do passo.
- **Inversão da string:** O uso de `[::-1]` inverte a ordem dos caracteres na string.
- **Métodos de formatação de string:** `.upper()` converte a string para maiúsculas, `.title()` converte a string para o estilo de título e `.lower()` converte a string para minúsculas.
- Estas são operações comuns em Python para manipulação de strings, permitindo diversas transformações e extrações de partes específicas das strings.

In [3]:
### Como usar o format para criar formatações personalizadas em prints e textos.
'''
:<		Alinha o texto à esquerda (se tiver espaço na tela para isso)
:>		Alinha o texto à direita (se tiver espaço na tela para isso)
:^		Alinha o texto ao centro (se tiver espaço na tela para isso)
:+		Coloca o sinal sempre na frente do número (independente se é positivo ou negativo)
:,		Coloca a vírgula como separador de milhar
:_		Coloca o _ como separador de milhar
:e		Formato Científico
:f		Número com quantidade fixa de casas decimais
:x		Formato HEX minúscula (para cores)
:X		Formato HEX maiúscula (para cores)
:%		Formato Percentual'''

'\n:<\t\tAlinha o texto à esquerda (se tiver espaço na tela para isso)\n:>\t\tAlinha o texto à direita (se tiver espaço na tela para isso)\n:^\t\tAlinha o texto ao centro (se tiver espaço na tela para isso)\n:+\t\tColoca o sinal sempre na frente do número (independente se é positivo ou negativo)\n:,\t\tColoca a vírgula como separador de milhar\n:_\t\tColoca o _ como separador de milhar\n:e\t\tFormato Científico\n:f\t\tNúmero com quantidade fixa de casas decimais\n:x\t\tFormato HEX minúscula (para cores)\n:X\t\tFormato HEX maiúscula (para cores)\n:%\t\tFormato Percentual'

O método `format()` em Python é usado para formatar strings de maneira flexível e personalizada. Aqui está um exemplo de como usar o método `format()` para criar formatações personalizadas em textos e impressões:

```python
# Exemplo de formatação personalizada usando o método format()
numero = 1234567.89
print("Número formatado com vírgula como separador de milhar: {:,}".format(numero))
print("Número formatado com underscore como separador de milhar: {:_}".format(numero))
print("Número formatado com sinal na frente: {:+}".format(numero))
print("Número formatado em formato científico: {:e}".format(numero))
print("Número formatado com duas casas decimais: {:.2f}".format(numero))
print("Número formatado em formato hexadecimal (minúsculas): {:x}".format(int(numero)))
print("Número formatado em formato hexadecimal (maiúsculas): {:X}".format(int(numero)))
print("Número formatado como percentual: {:.2%}".format(0.25))
```

**Saída:**

```
Número formatado com vírgula como separador de milhar: 1,234,567.89
Número formatado com underscore como separador de milhar: 1_234_567.89
Número formatado com sinal na frente: +1234567.89
Número formatado em formato científico: 1.234568e+06
Número formatado com duas casas decimais: 1234567.89
Número formatado em formato hexadecimal (minúsculas): 12d687
Número formatado em formato hexadecimal (maiúsculas): 12D687
Número formatado como percentual: 25.00%
```

Esses são apenas alguns exemplos de como você pode usar o método `format()` para criar formatações personalizadas em suas strings. Ele oferece uma maneira poderosa e flexível de controlar a aparência de seus dados e textos de saída em Python.

## **try & except**

In [24]:
#Exemplo
numero = input('Digite um numero: ')

try:
    num = int(numero)
    dobro = num * 2
    
    print(f'O numero digitado foi {num} seu dobro e {dobro}')
    
except:
    print('Erro!')

O numero digitado foi 45 seu dobro e 90


Neste exemplo, o programa solicita ao usuário que insira um número. Em seguida, ele tenta converter esse número para um inteiro usando a função `int()`. Se a conversão for bem-sucedida, o programa calcula o dobro desse número e o imprime na tela. Se ocorrer algum erro durante a conversão, como quando o usuário digita algo que não pode ser convertido para um número inteiro, o bloco `except` será executado e uma mensagem de erro será exibida.

```python
# Solicita ao usuário que insira um número
numero = input('Digite um numero: ')

try:
    # Tenta converter o número para inteiro
    num = int(numero)
    
    # Calcula o dobro do número
    dobro = num * 2
    
    # Imprime o número digitado e seu dobro
    print(f'O número digitado foi {num} e seu dobro é {dobro}')
    
except:
    # Se ocorrer um erro durante a conversão, imprime uma mensagem de erro
    print('Erro!')
```

Este código demonstra como usar a declaração `try-except` em Python para lidar com possíveis erros durante a execução do programa, garantindo que o programa não seja interrompido abruptamente se ocorrerem exceções.

In [25]:
# Exemplo
primeiro = input('Primeiro numero: ')
segundo = input('Segundo numero: ')

try:
    p = float(primeiro)
    s = float(segundo)
    
    soma = p + s
    print(f'A soma dos numeros e {soma}')
    
except:
    print(ValueError)

A soma dos numeros e 90.0


In [6]:
#Contantes e Variaveis
CONSTANTE = 'Programando em Python' #Constante e uma variavel que não muda sempre sera toda em Maiuscula
variavel  = 'Programa em Python' #Variavel que muda

print(f"A constante é {CONSTANTE} {type(CONSTANTE)}")#Imprimindo a Constante
print(f'A variavel e {variavel} {type(variavel)}')

A constante é Programando em Python <class 'str'>
A variavel e Programa em Python <class 'str'>




```python
# Declaração de uma constante
CONSTANTE = 'Programando em Python'  # Constante é uma variável que não muda, geralmente escrita toda em maiúsculas

# Declaração de uma variável
variavel = 'Programa em Python'  # Variável é uma entidade que pode armazenar valores e pode mudar ao longo do tempo

# Impressão da constante
print(f"A constante é {CONSTANTE} {type(CONSTANTE)}")

# Impressão da variável
print(f'A variável é {variavel} {type(variavel)}')
```

**Explicação:**

- Uma constante em Python é uma variável que não muda de valor durante a execução do programa. Por convenção, o nome de uma constante é todo em maiúsculas.
- Uma variável em Python é uma entidade que pode armazenar valores e pode mudar ao longo do tempo.
- O comando `print()` é usado para imprimir o valor das variáveis `CONSTANTE` e `variavel`, bem como seus tipos, utilizando a formatação de string f-string (`f"{variavel}"`).

# Função Print

In [5]:
#Sep = separador
#end = final do print, ele uni as linhas
print(12, 13, 45, sep=' - ')
print('Meu nome e ', end='')
print('Francisco')

12 - 13 - 45
Meu nome e Francisco


Explicação

```python
# Imprime os números 12, 13 e 45, separados por um hífen e espaço
print(12, 13, 45, sep=' - ')

# Imprime 'Meu nome e ' sem pular para a próxima linha
print('Meu nome e ', end='')

# Imprime 'Francisco' na mesma linha da string anterior
print('Francisco')
```

**Explicação:**

- O parâmetro `sep` é usado para especificar o separador entre os argumentos passados para a função `print()`. Aqui, você define o separador como `' - '`, o que resulta em todos os números impressos separados por um hífen e um espaço.
- O parâmetro `end` é usado para especificar o que deve ser impresso no final da função `print()`. No primeiro `print()`, você não especificou o `end`, portanto, a função `print()` automaticamente adiciona uma nova linha. No segundo `print()`, você definiu `end=''`, o que significa que nada será impresso no final da primeira string, permitindo que a próxima string seja impressa na mesma linha.

# **Tipos de Erros no Python**

**SyntaxError**: Erro de sintaxe, geralmente ocorre quando há um erro na estrutura do código Python.

**IndentationError**: Ocorre quando a indentação do código não está correta, especialmente em blocos como loops e funções.

**NameError**: Indica que uma variável ou função usada não foi definida.

**TypeError**: Surge quando uma operação é aplicada a um tipo incompatível de dado.

**ValueError**: Ocorre quando uma função recebe um argumento com o tipo certo, mas valor inapropriado.

**IndexError**: Acontece quando um índice usado para acessar um elemento de uma sequência está fora do intervalo.

**KeyError**: Ocasionado ao tentar acessar uma chave que não existe em um dicionário.

**AttributeError**: Surge quando uma referência a um atributo ou método de um objeto é feita, mas o objeto não possui esse atributo ou método.

**ImportError**: Acontece quando uma instrução import ou from import falha em importar um módulo.

**ModuleNotFoundError**: Similar ao ImportError, mas mais específico, indicando que um módulo importado não pode ser encontrado.

**FileNotFoundError**: Ocorre quando um arquivo ou diretório não pode ser encontrado.

**IOError**: Indica um erro de entrada/saída, geralmente relacionado à operações de leitura ou escrita em arquivos.

**PermissionError**: Surge quando não há permissão para realizar uma operação em um arquivo ou diretório.

**ZeroDivisionError**: Acontece quando uma divisão é feita por zero.

**OverflowError**: Ocorre quando um cálculo resulta em um valor que é muito grande para ser representado.

**MemoryError**: Indica que o Python não possui memória suficiente para executar uma operação.

**RecursionError**: Ocorre quando uma função chama a si mesma recursivamente de forma excessiva.

**KeyboardInterrupt**: Surge quando o usuário interrompe a execução de um programa pressionando Ctrl + C.

**SystemExit**: Indica uma solicitação para sair do interpretador Python.

**AssertionError**: Ocorre quando uma instrução assert falha.

**TypeError**: Indica que um objeto não é iterável, mas uma operação de iteração é tentada.

**UnboundLocalError**: Acontece quando uma variável local é referenciada antes de ser atribuída.

**NotImplementedError**: Indica que uma funcionalidade ainda não foi implementada.

**DeprecationWarning**: Aviso indicando que uma funcionalidade está obsoleta e será removida em versões futuras do Python.

**RuntimeError**: Um erro genérico que ocorre durante a execução do programa, geralmente indicando um problema imprevisto.

# Comentarios em Python

In [26]:
# Isso é um comentário

# Isso é um comentário (mas como é uma string não atribuída a uma variável ou usada, é considerado um comentário)
'Isso é Um Comentario'
"Isso é Um Comentario"

"""
Isso é um comentário
em várias linhas
"""

# Tudo que estiver entre aspas fora da variável é considerado um comentário
"Tudo que estiver entre aspas fora da variável é considerado um comentário"


'Tudo que estiver entre aspas fora da variável é considerado um comentário'